src/main/java/io/github/fablabsmc/fablabs/api/fluidvolume/v1/volume/api/FluidVolume.java

```
@@ -0,0 +1,223 @@
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Objects;
+
+import com.google.common.collect.Iterators;
+import io.github.fablabsmc.fablabs.api.fluidvolume.v1.math.Fraction;
+import io.github.fablabsmc.fablabs.api.fluidvolume.v1.properties.FluidPropertyManager;
+
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+
+/**
+ * a mutable representation of a quantity and it's data of a fluid.
+ * this class can and should be extended for custom logic when dealing with volume merging.
+ *
+ * @see #merge(FluidVolume)
+ * @see #drain(FluidVolume)
+ * @see #draw(Fraction)
+ * @see FixedSizedFluidVolume
+ */
+public class FluidVolume extends AbstractCollection<FluidContainer> implements FluidContainer, SingleFluidContainer {
+	// not nullable
+	protected Fluid fluid;
+	// not nullable
+	protected Fraction amount;
+	// not nullable
+	protected CompoundTag tag;
+
+	/**
+	 * creates a new empty fluid container.
+	 */
+	public FluidVolume() {
+		this(Fluids.EMPTY, Fraction.ZERO);
+	}
+
+	/**
+	 * creates a new fluid container with an empty data tag.
+	 *
+	 * @param fluid  the fluid
+	 * @param amount the amount of fluid contained, must be 0 if fluid is null
+	 */
+	public FluidVolume(Fluid fluid, Fraction amount) {
+		this(fluid, amount, new CompoundTag());
+	}
+
+	/**
+	 * creates a new fluid container with custom data.
+	 *
+	 * @param fluid  the fluid
+	 * @param amount the amount of fluid contained, must be 0 if fluid is EMPTY
+	 * @param tag    the data
+	 */
+	public FluidVolume(Fluid fluid, Fraction amount, CompoundTag tag) {
+		if (fluid == Fluids.EMPTY && amount.getNumerator() != 0) {
+			throw new IllegalArgumentException("EMPTY fluid must have zero amount!");
+		}
+
+		if (fluid == null) {
+			throw new IllegalArgumentException("Fluid must not be null! Hint: Fluids.EMPTY");
```

I'd say null check at head of method so you can fail fast.